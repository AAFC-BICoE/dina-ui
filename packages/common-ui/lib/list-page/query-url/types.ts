export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
