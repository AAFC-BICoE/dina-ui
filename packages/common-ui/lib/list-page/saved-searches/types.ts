import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
