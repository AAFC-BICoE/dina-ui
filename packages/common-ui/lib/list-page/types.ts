import { KitsuResource } from "kitsu";
import { Column } from "react-table";

/**
 * This type extends the react-table column type, this just adds a few specific fields for elastic
 * search mapping and internationalization.
 */
export interface TableColumn<TData extends KitsuResource>
  extends Column<TData> {
  /**
   * User-friendly column to be displayed. You can use a DinaMessage key for internationalization.
   */
  label?: string;

  /**
   * Elastic search path to the attribute.
   *
   * Example: `data.attributes.name`
   */
  attributePath?: string;

  /**
   * This field is used to find the relationship in the included section.
   */
  relationshipType?: string;

  /**
   * Is this attribute considered a keyword in elastic search. Required for filtering and sorting.
   */
  isKeyword?: boolean;

  /**
   * The QueryPage will only display the accessors that are displayed on the result table. However,
   * if you have custom cells that receive other fields you will need to add them to this list so
   * elastic search includes the fields in the result.
   *
   * Example: `data.attributes.name`
   *
   * Please note that duplicate fields are automatically removed so you don't need to worry about
   * having unique accessors.
   */
  additionalAccessors?: string[];
}

/**
 * The full path will be generated for elastic using a combination of the parent path and
 * the value. The path is generated using the following:
 *
 * {parentPath}.{path}.{value}
 *
 * Example: included.attributes.determination.verbatimScientificName
 */
export interface ESIndexMapping {
  /**
   * Name of the attribute.
   *
   * Example: verbatimScientificName
   */
  value: string;

  /**
   * Text that is displayed to the user in the Query Filtering option menu.
   *
   * This text is a user-friendly generated label, which may show some paths to help the user
   * understand the relationships better. This is generated from the path.
   *
   * Example: determination.verbatimScientificName
   */
  label: string;

  /**
   * The attributes type. This can change how the query row is displayed and the options provided.
   *
   * Examples: text, keyword, boolean, date, boolean, long, short, integer...
   */
  type: string;

  /**
   * If enabled, it will allow the user to see suggestions as they type. The suggestions will come
   * from elastic search based on most common values saved.
   *
   * Only available for the text type.
   */
  distinctTerm: boolean;

  /**
   * If enabled it will allow the user to search based on the starting of a word.
   *
   * Example: Hexapoda can be matched with "hex".
   */
  startsWithSupport: boolean;

  /**
   * If enabled it will allow the user to search based in the middle of a word.
   *
   * Example: Hexapoda can be matched with "pod".
   */
  containsSupport: boolean;

  /**
   * If enabled it will allow the user to search based on the ending of a word.
   *
   * Example: Hexapoda can be matched with "poda".
   */
  endsWithSupport: boolean;

  /**
   * The path for the attribute without the attribute name. This path does not include the parent
   * path.
   *
   * Example: attribute.determination
   */
  path: string;

  /**
   * If the attribute belongs to a relationship, this is the path for only the parent. When generating
   * the elastic search query it will use this as the prefix of the path.
   *
   * Example: included
   */
  parentPath?: string;

  /**
   * If the attribute belongs to a relationship, this is the name which will be used to group
   * attributes under the same relationship together in the search. This name will also be used to
   * display text of the group.
   *
   * This text should match the user-friendly label in the locales files. It will be searched using
   * title_[parentName].
   *
   * Example: preparationMethod (and the label will be `title_preparationMethod`)
   */
  parentName?: string;

  /**
   * The parent type is the relationship type to be used. This will be used for elastic search
   * filtering. The reason this is used since the same field can be used for multiple relationships.
   *
   * Example: preparation-method
   */
  parentType?: string;

  /**
   * Only provided if it was added using a dynamic field config.
   */
  dynamicField?: DynamicField;
}

/**
 * Each type on the query builder has a function that is used to transform the query builder row
 * into elastic search logic.
 *
 * @see QueryBuilderElasticSearchExport
 * @see QueryRowTextSearch
 */
export interface TransformToDSLProps {
  /**
   * The query builder type, not really used but required in the function for the QueryBuilder
   * library.
   *
   * For the Elastic Search type you can use the fieldInfo.type instead.
   */
  queryType: string;

  /**
   * The value to search against. If the type is "text" then this is the text that will be filtered.
   */
  value: string;

  /**
   * The operation being performed.
   *
   * For example:
   * "equals", "notEquals", "containsText"
   *
   * Operators are defined in the QueryBuilderConfig file.
   */
  operation: string;

  /**
   * This field path is the unique path given to each item in the field list.
   *
   * For getting the proper absolute field path, use the FieldInfo to generate it.
   */
  fieldPath: string;

  /**
   * The elastic search mapping for the field.
   */
  fieldInfo?: ESIndexMapping;
}

export type DynamicFieldType = "managedAttribute" | "fieldExtension";

export interface DynamicFieldsMappingConfig {
  /** Attribute level dynamic fields */
  fields: DynamicField[];

  /** Dynamic fields for relationships */
  relationshipFields: RelationshipDynamicField[];
}

export interface DynamicField {
  /**
   * Option label that should be used.
   */
  label: string;

  type: DynamicFieldType;

  path: string;

  /**
   * Endpoint where these dynamic fields can be retrieved to list.
   *
   * Example: "collection-api/managed-attribute"
   */
  apiEndpoint: string;

  /**
   * Optional field to indicate which Managed Attributes or Field Extensions should be listed.
   */
  component?: string;
}

/**
 * Configuration for where the Dynamic Field can be found within the relationship index mapping.
 */
export interface RelationshipDynamicField extends DynamicField {
  referencedBy: string;
  referencedType: string;
}
