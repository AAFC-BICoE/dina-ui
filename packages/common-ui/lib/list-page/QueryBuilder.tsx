import {
  DinaFormSection,
  FieldWrapperProps,
  LoadingSpinner,
  SelectField,
  useApiClient
} from "..";
import { QueryRow } from "./QueryRow";
import { FieldArray } from "formik";
import { GroupSelectField } from "../../../dina-ui/components";
import { useEffect, useMemo } from "react";
import { v4 as uuidv4 } from "uuid";
import useSWR from "swr";
import { ESIndexMapping } from "./types";

import {
  Query,
  Builder,
  Utils as QbUtils,
  BasicConfig
} from "react-awesome-query-builder";
import { useState, useCallback } from "react";
import {
  JsonGroup,
  Config,
  ImmutableTree,
  BuilderProps
} from "react-awesome-query-builder";
import { Button } from "react-bootstrap";
import { FaTrash } from "react-icons/fa";
import { QueryFieldSelector } from "./QueryFieldSelector";

interface QueryBuilderProps extends FieldWrapperProps {
  indexName: string;
  onGroupChange: (currentSubmittedValues: any) => void;
}

export function QueryBuilder({
  name,
  indexName,
  onGroupChange
}: QueryBuilderProps) {
  const { apiClient } = useApiClient();

  // State to store the index map after it has been retrieved.
  const [indexMap, setIndexMap] = useState<ESIndexMapping[]>([]);

  // Query Builder Configuration, only needs to update if the index map has changed.
  const config: Config = useMemo(
    () => ({
      conjunctions: {
        ...BasicConfig.conjunctions
      },
      operators: {
        equals: {
          label: "equals"
        }
      },
      widgets: {
        ...BasicConfig.widgets
      },
      types: {
        ...BasicConfig.types
      },
      fields: {
        test: {
          label: "test",
          type: "number"
        }
      },
      settings: {
        ...BasicConfig.settings,
        renderButton: (buttonProps) => {
          if (buttonProps) {
            switch (buttonProps?.type) {
              case "addRule":
              case "addGroup":
                return (
                  <Button onClick={buttonProps?.onClick} className="ms-1">
                    {buttonProps.label}
                  </Button>
                );
              case "delGroup":
              case "delRule":
              case "delRuleGroup":
                return (
                  <Button
                    onClick={buttonProps?.onClick}
                    className="ms-1"
                    variant="danger"
                  >
                    <FaTrash />
                  </Button>
                );
            }
          }

          return (
            <Button onClick={buttonProps?.onClick} className="ms-1">
              {buttonProps?.label}
            </Button>
          );
        },
        renderField: (fieldDropdownProps) => (
          <QueryFieldSelector indexMap={indexMap} />
        ),
        showNot: false,
        canRegroup: true,
        canReorder: true
      }
    }),
    [indexMap]
  );

  // State to store the query tree generated by the Query Builder.
  const [queryTree, setQueryTree] = useState<ImmutableTree>(
    QbUtils.checkTree(
      QbUtils.loadTree({
        id: QbUtils.uuid(),
        type: "group"
      }),
      config
    )
  );

  const onChange = useCallback((immutableTree: ImmutableTree) => {
    setQueryTree(immutableTree);
  }, []);

  const renderBuilder = useCallback(
    (props: BuilderProps) => (
      <div className="query-builder-container">
        <div className="query-builder qb-lite">
          <Builder {...props} />
        </div>
      </div>
    ),
    []
  );

  // Retrieve the index mapping.
  useEffect(() => {
    async function getIndexMapping() {
      const mapping = await fetchQueryFieldsByIndex(indexName);
      setIndexMap(mapping);
    }
    getIndexMapping();
  }, []);

  /**
   * The query builder options are generated from the elastic search index. This method will
   * request the mappings from the index.
   *
   * @param searchIndexName index to retrieve from.
   * @returns ESIndexMapping[]
   */
  async function fetchQueryFieldsByIndex(searchIndexName) {
    const resp = await apiClient.axios.get("search-api/search-ws/mapping", {
      params: { indexName: searchIndexName }
    });

    const result: ESIndexMapping[] = [];

    // Read index attributes.
    resp.data?.attributes
      ?.filter((key) => key.name !== "type")
      .map((key) => {
        const path = key.path;
        const prefix = "data.attributes";
        let attrPrefix;
        if (path && path.includes(prefix)) {
          attrPrefix = path.substring(prefix.length + 1);
        }
        result.push({
          label: attrPrefix ? attrPrefix + "." + key.name : key.name,
          value: key.path
            ? key.path + "." + key.name
            : key.name === "id"
            ? "data." + key.name
            : key.name,
          type: key.type,
          path: key.path,
          distinctTerm: key.distinct_term_agg
        });
      });

    // Read relationship attributes.
    resp.data?.relationships?.map((relationship) => {
      relationship?.attributes?.map((relationshipAttribute) => {
        // This is the user-friendly label to display on the search dropdown.
        const attributeLabel = relationshipAttribute.path?.includes(".")
          ? relationshipAttribute.path.substring(
              relationshipAttribute.path.indexOf(".") + 1
            ) +
            "." +
            relationshipAttribute.name
          : relationshipAttribute.name;

        result.push({
          label: attributeLabel,
          value: relationship.value + "." + attributeLabel,
          type: relationshipAttribute.type,
          path: relationshipAttribute.path,
          parentName: relationship.referencedBy,
          parentType: relationship.value,
          parentPath: relationship.path,
          distinctTerm: relationshipAttribute.distinct_term_agg
        });
      });
    });
    return result;
  }

  // Display loading spinner when performing request for the index.
  if (!indexMap) {
    return <LoadingSpinner loading={true} />;
  }

  const sortedData = indexMap
    ?.sort((a, b) => a.label.localeCompare(b.label))
    .filter((prop) => !prop.label.startsWith("group"));

  return (
    <>
      <Query
        {...config}
        value={queryTree}
        onChange={onChange}
        renderBuilder={renderBuilder}
      />
      <FieldArray name={name}>
        {(fieldArrayProps) => {
          const elements: [] = fieldArrayProps.form.values.queryRows;

          function addRow() {
            fieldArrayProps.push(
              <QueryRow
                name={fieldArrayProps.name}
                indexName={indexName}
                esIndexMapping={sortedData as any}
                index={elements?.length ?? 0}
                removeRow={removeRow}
                addRow={addRow}
              />
            );
            // initialize the logic switch value to be "and"
            fieldArrayProps.form.setFieldValue(
              `${fieldArrayProps.name}[${
                elements?.length ?? 0
              }].compoundQueryType`,
              "and"
            );
          }

          function removeRow(index) {
            fieldArrayProps.remove(index);
          }

          return elements?.length > 0
            ? elements?.map((_, index) => (
                <QueryRow
                  name={fieldArrayProps.name}
                  indexName={indexName}
                  key={index}
                  index={index}
                  addRow={addRow}
                  removeRow={removeRow}
                  esIndexMapping={sortedData as any}
                />
              ))
            : null;
        }}
      </FieldArray>
      <DinaFormSection horizontal={"flex"}>
        <GroupSelectField
          isMulti={true}
          name="group"
          className="col-md-4"
          onChange={(value, formik) =>
            onGroupChange({
              submittedValues: { ...formik.values, group: value }
            })
          }
        />
      </DinaFormSection>
    </>
  );
}
