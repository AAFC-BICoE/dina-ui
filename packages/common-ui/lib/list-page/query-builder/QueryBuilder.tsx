import { LoadingSpinner } from "../..";
import { Query, Builder, Utils, JsonTree } from "react-awesome-query-builder";
import { useCallback } from "react";
import {
  Config,
  ImmutableTree,
  BuilderProps
} from "react-awesome-query-builder";
import React from "react";
import { Button } from "react-bootstrap";
import { SavedSearch } from "../saved-searches/SavedSearch";
import { DinaMessage } from "../../../../dina-ui/intl/dina-ui-intl";

interface QueryBuilderProps {
  /**
   * Index name being used for the QueryPage.
   */
  indexName: string;

  /**
   * Query Builder Configuration.
   */
  queryBuilderConfig?: Config;

  /**
   * Query Builder local tree.
   */
  queryBuilderTree?: ImmutableTree;

  /**
   * Callback function to indicate that the query builder has changed.
   */
  setQueryBuilderTree: (newTree: ImmutableTree) => void;

  /**
   * Callback function to indicate the form was submitted.
   */
  onSubmit: () => void;

  /**
   * Callback to indicate the query builder was reset.
   */
  onReset: () => void;
}

function QueryBuilder({
  indexName,
  queryBuilderConfig,
  queryBuilderTree,
  setQueryBuilderTree,
  onSubmit,
  onReset
}: QueryBuilderProps) {
  const onChange = useCallback((immutableTree: ImmutableTree) => {
    setQueryBuilderTree(immutableTree);
  }, []);

  const renderBuilder = useCallback(
    (props: BuilderProps) => (
      <div className="query-builder-container">
        <div className="query-builder qb-lite">
          <Builder {...props} />
        </div>
      </div>
    ),
    []
  );

  // Display loading spinner when performing request for the index.
  if (!queryBuilderTree || !queryBuilderConfig) {
    return <LoadingSpinner loading={true} />;
  }

  return (
    <>
      <label
        style={{
          fontSize: 20,
          fontFamily: "sans-serif",
          fontWeight: "bold"
        }}
        className="mb-2"
      >
        <DinaMessage id="search" />
      </label>
      <SavedSearch
        indexName={indexName}
        queryBuilderTree={queryBuilderTree}
        setQueryBuilderTree={setQueryBuilderTree}
        queryBuilderConfig={queryBuilderConfig}
        performSubmit={onSubmit}
      />
      <Query
        {...queryBuilderConfig}
        value={queryBuilderTree}
        onChange={onChange}
        renderBuilder={renderBuilder}
      />
      <div className="mt-2">
        <Button onClick={onSubmit} className="me-2">
          Search
        </Button>
        <Button onClick={onReset} variant="secondary">
          Reset
        </Button>
      </div>
    </>
  );
}

/**
 * Default query tree contains an empty rule so it's one less the step the user needs to perform.
 *
 * This is the default when first loading the page and resetting the query builder.
 */
export function defaultQueryTree(): ImmutableTree {
  const groupId = "8c6dc2c8-4070-48ce-b700-13a931f9ebaf";
  const ruleId = "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6";

  return Utils.loadTree({
    id: groupId,
    type: "group",
    children1: {
      "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6": {
        type: "rule",
        id: ruleId,
        properties: {
          field: null,
          operator: null,
          value: [],
          valueSrc: [],
          valueError: []
        },
        path: [groupId, ruleId]
      }
    },
    properties: { conjunction: "AND" },
    path: [groupId]
  } as JsonTree);
}

/**
 * Generate an empty tree with no rules. Used for special cases like custom elastic search
 * view queries where the query builder is not needed.
 */
export function emptyQueryTree(): ImmutableTree {
  return Utils.loadTree({
    id: Utils.uuid(),
    type: "group",
    children1: {}
  } as JsonTree);
}

/**
 * Currently utilized for the custom view queries to show associated records.
 *
 * @param uuid UUID to search for.
 * @param path Relationship path to search against.
 * @returns QueryBuilder tree which will be translated into elastic search query.
 */
export function generateUUIDTree(uuid: string, path: string): JsonTree {
  return {
    id: Utils.uuid(),
    type: "group",
    children1: {
      [Utils.uuid()]: {
        type: "rule",
        properties: {
          field: path,
          operator: "uuid",
          value: [uuid]
        }
      }
    }
  };
}

export const QueryBuilderMemo = React.memo(QueryBuilder);
