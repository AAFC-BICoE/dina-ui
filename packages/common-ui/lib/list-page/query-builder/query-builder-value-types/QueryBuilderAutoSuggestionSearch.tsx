import { ESIndexMapping } from "../../types";
import { useElasticSearchDistinctTerm } from "../../useElasticSearchDistinctTerm";
import AutoSuggest, { InputProps } from "react-autosuggest";
import React, { useEffect, useMemo, useState } from "react";
import { useIntl } from "react-intl";
import { noop } from "lodash";
import { useAccount } from "common-ui/lib";

interface QueryBuilderAutoSuggestionTextSearchProps {
  /**
   * Current field selected in the Query Builder.
   */
  currentFieldName?: string;

  /**
   * Current match type being used.
   */
  matchType?: string;

  /**
   * Retrieve the current value from the Query Builder.
   */
  value: string;

  /**
   * Pass the selected value to the Query Builder to store.
   */
  setValue: ((fieldPath: string) => void) | undefined;

  /**
   * Required in order to know what index to search against elastic search for the suggestions.
   */
  indexName: string;

  /**
   * The elastic search mapping changes based on the field that is selected. This is used to
   * create the field name for the elastic search request.
   */
  indexMap?: ESIndexMapping[];
}

function QueryBuilderAutoSuggestionTextSearch({
  currentFieldName,
  matchType,
  value,
  setValue,
  indexName,
  indexMap
}: QueryBuilderAutoSuggestionTextSearchProps) {
  const { formatMessage } = useIntl();
  const { groupNames } = useAccount();

  // Index settings for the field currently selected.
  const [fieldSettings, setFieldSettings] = useState<
    ESIndexMapping | undefined
  >();

  // Field name for the field currently selected.
  const [fieldName, setFieldName] = useState<string>();

  // State to store if the text field is focused. Used to determine if suggestions should be
  // displayed.
  const [focus, setFocus] = useState<boolean>(false);

  // Retrieve the suggestions using elastic search. Only updates if the field/group change.
  const suggestions = useElasticSearchDistinctTerm({
    fieldName,
    groups: groupNames ?? [],
    relationshipType: fieldSettings?.parentType,
    indexName
  });

  // Whenever the current field name changes, retrieve the new field settings for that field.
  useEffect(() => {
    setFieldSettings(
      indexMap?.find((attribute) => attribute.value === currentFieldName)
    );
  }, [currentFieldName]);

  useEffect(() => {
    if (!fieldSettings) return;

    setFieldName(
      fieldSettings?.parentPath
        ? fieldSettings?.parentPath +
            "." +
            fieldSettings?.path +
            "." +
            fieldSettings?.label
        : fieldSettings?.value
    );
  }, [fieldSettings]);

  const inputProps: InputProps<any> = {
    placeholder: formatMessage({ id: "typeHereToSearch" }) ?? "",
    value: value ?? "",
    onChange: (_event, { newValue }) => {
      setValue?.(newValue);
    },
    autoComplete: "none",
    className: "form-control",
    onFocus: () => setFocus(true),
    onBlur: () => setFocus(false)
  };

  // Filter the suggestion list based on the value.
  const currentSuggestions = useMemo(
    () =>
      suggestions?.filter(
        (suggestion) =>
          suggestion?.toLowerCase()?.includes(value?.toLowerCase() ?? "") &&
          suggestion !== value
      ),
    [value, suggestions]
  );

  return (
    <>
      {/* Depending on the matchType, it changes the rest of the query row. */}
      {(matchType === "equals" || matchType === "notEquals") && (
        <>
          <style>{`
          .autosuggest-container {
            position: relative;
            width: 100%;
          }
          .autosuggest .suggestions-container {
            display: none;
          }
          .autosuggest .suggestions-container-open {
            display: block;
            position: absolute;
            width: 100%;
            z-index: 20;
          }
          .autosuggest .suggestion-highlighted { 
            background-color: #ddd;
            cursor: pointer;
          }
        `}</style>
          <div className="autosuggest">
            <AutoSuggest<string>
              multiSection={false}
              suggestions={currentSuggestions}
              getSuggestionValue={(s) => s}
              onSuggestionsFetchRequested={({ value: fetchValue }) =>
                setValue?.(fetchValue)
              }
              onSuggestionSelected={(_event, data) =>
                setValue?.(data.suggestion)
              }
              onSuggestionsClearRequested={noop}
              renderSuggestion={(text) => <div>{text}</div>}
              inputProps={inputProps}
              alwaysRenderSuggestions={focus}
              theme={{
                suggestionsList: "list-group",
                suggestion: "list-group-item",
                suggestionHighlighted: "suggestion-highlighted",
                suggestionsContainerOpen: "suggestions-container-open",
                suggestionsContainer: "suggestions-container",
                container: "autosuggest-container"
              }}
            />
          </div>
        </>
      )}
    </>
  );
}

export const QueryBuilderAutoSuggestionTextSearchMemo = React.memo(
  QueryBuilderAutoSuggestionTextSearch
);
