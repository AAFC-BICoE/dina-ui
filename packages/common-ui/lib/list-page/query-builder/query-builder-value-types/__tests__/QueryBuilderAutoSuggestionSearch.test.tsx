import { mountWithAppContext2 } from "common-ui/lib/test-util/mock-app-context";
import { QueryBuilderAutoSuggestionTextSearchMemo } from "../QueryBuilderAutoSuggestionSearch";
import { waitFor } from "@testing-library/react";
import { DinaForm } from "common-ui/lib/formik-connected/DinaForm";
import { ESIndexMapping } from "../../../types";
import { QueryBuilderContextProvider } from "../../QueryBuilder";
import { noop } from "lodash";
import userEvent from "@testing-library/user-event";

const INDEX_NAME = "dina-material-sample-index";

const CURRENT_FIELD_NAME = "data.attributes.materialSampleType";

const INDEX_MAP: ESIndexMapping[] = [
  {
    label: "materialSampleType",
    value: "data.attributes.materialSampleType",
    type: "text",
    path: "data.attributes",
    distinctTerm: true,
    keywordMultiFieldSupport: true,
    optimizedPrefix: false,
    containsSupport: false,
    endsWithSupport: false,
    hideField: false,
    keywordNumericSupport: false
  }
];

const mockAutoSuggestionRequest = jest.fn<any, any>(async (path) => {
  if (path === "search-api/search-ws/search") {
    return {
      data: {
        took: 3,
        timed_out: false,
        _shards: {
          failed: 0.0,
          successful: 1.0,
          total: 1.0,
          skipped: 0.0
        },
        hits: {
          total: {
            relation: "eq",
            value: 0
          },
          hits: []
        },
        aggregations: {
          "sterms#term_aggregation": {
            buckets: [
              {
                doc_count: 7,
                key: "WHOLE_ORGANISM"
              },
              {
                doc_count: 6,
                key: "MIXED_ORGANISMS"
              },
              {
                doc_count: 5,
                key: "MOLECULAR_SAMPLE"
              },
              {
                doc_count: 1,
                key: "CULTURE_STRAIN"
              }
            ],
            doc_count_error_upper_bound: 0,
            sum_other_doc_count: 0
          }
        }
      }
    };
  }
});

const apiClientMock = {
  apiClient: {
    axios: { post: mockAutoSuggestionRequest } as any
  }
};

describe("QueryBuilderAutoSuggestionSearch", () => {
  beforeEach(jest.clearAllMocks);

  describe("QueryBuilderAutoSuggestionSearch Component", () => {
    it("Display field if match type is equals", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const autoSuggestionEquals = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider value={{ performSubmit: noop }}>
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="equals"
              value="test"
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        { apiContext: apiClientMock }
      );

      // Expect a snapshot with the date field being displayed.
      expect(autoSuggestionEquals.asFragment()).toMatchSnapshot(
        "Expect date field to be displayed since match type is equals"
      );

      const autoSuggestionEmpty = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider value={{ performSubmit: noop }}>
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="empty"
              value="test"
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        { apiContext: apiClientMock }
      );

      // Expect a snapshot without the date field being displayed.
      expect(autoSuggestionEmpty.asFragment()).toMatchSnapshot(
        "Expect date field not to be displayed since the match type is not equals"
      );
    });

    it("Display different placeholder on IN operators", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const autoSuggestionIn = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider value={{ performSubmit: noop }}>
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="in"
              value="test"
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        { apiContext: apiClientMock }
      );

      // Expect a snapshot with a specific placeholder.
      expect(autoSuggestionIn.asFragment()).toMatchSnapshot(
        "Expect auto-suggestion field to be displayed with a different placeholder."
      );

      const autoSuggestionNotIn = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider value={{ performSubmit: noop }}>
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="notIn"
              value="test"
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        { apiContext: apiClientMock }
      );

      // Expect a snapshot with a specific placeholder.
      expect(autoSuggestionNotIn.asFragment()).toMatchSnapshot(
        "Expect auto-suggestion field to be displayed with a different placeholder."
      );
    });

    it("Display suggestions on equals or not equals operators", async () => {
      const autoSuggestionComponent = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider value={{ performSubmit: noop }}>
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="equals"
              value=""
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        { apiContext: apiClientMock }
      );

      // Wait for the auto-suggestion element to be loaded in.
      await waitFor(() =>
        expect(autoSuggestionComponent.findByRole("textbox")).not.toBeNull()
      );

      // Wait for the api call to be made.
      await waitFor(() => expect(mockAutoSuggestionRequest).toBeCalledTimes(1));

      // Ensure the API request was made correctly.
      expect(mockAutoSuggestionRequest).toBeCalledWith(
        "search-api/search-ws/search", // Search endpoint
        {
          aggs: {
            term_aggregation: {
              terms: {
                field: "data.attributes.materialSampleType.keyword",
                size: 100
              }
            }
          },
          query: {
            terms: {
              "data.attributes.group": ["aafc", "cnc"]
            }
          },
          size: 0
        },
        { params: { indexName: "dina-material-sample-index" } }
      );

      // Simulate focus, to display the suggestions dialog.
      autoSuggestionComponent.getByRole("textbox").focus();

      // Wait for the component to update after focus change
      await waitFor(
        () => expect(autoSuggestionComponent.getByRole("textbox")).toHaveFocus
      );

      // Get each suggestion item and assert its content
      const suggestions = autoSuggestionComponent.getAllByRole("option");
      expect(suggestions.length).toBe(4); // Ensure there are 4 options

      // Assert the text content of each suggestion
      expect(suggestions[0].textContent).toBe("WHOLE_ORGANISM");
      expect(suggestions[1].textContent).toBe("MIXED_ORGANISMS");
      expect(suggestions[2].textContent).toBe("MOLECULAR_SAMPLE");
      expect(suggestions[3].textContent).toBe("CULTURE_STRAIN");
    });

    it("Should call performSubmit on enter key press in textfield", async () => {
      const mockPerformSubmit = jest.fn();
      const { getByRole } = mountWithAppContext2(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: mockPerformSubmit }}
          >
            <QueryBuilderAutoSuggestionTextSearchMemo
              indexName={INDEX_NAME}
              indexMap={INDEX_MAP}
              currentFieldName={CURRENT_FIELD_NAME}
              matchType="equals"
              value="test"
              setValue={jest.fn}
            />
          </QueryBuilderContextProvider>
        </DinaForm>
      );

      // Find the text field element
      const textField = getByRole("textbox");

      // Expect performSubmit to not be called yet.
      expect(mockPerformSubmit).toHaveBeenCalledTimes(0);

      // Simulate user typing "enter" key
      userEvent.type(textField, "{enter}");

      // Expect performSubmit to be called once
      expect(mockPerformSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
