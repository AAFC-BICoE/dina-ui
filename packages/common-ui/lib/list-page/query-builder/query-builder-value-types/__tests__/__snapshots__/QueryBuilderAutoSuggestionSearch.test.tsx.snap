// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderAutoSuggestionSearch QueryBuilderAutoSuggestionSearch Component Display different placeholder on IN operators: Expect auto-suggestion field to be displayed with a different placeholder. 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <input
        class="form-control"
        placeholder="queryBuilder_value_in_placeholder"
        type="text"
        value="test"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderAutoSuggestionSearch QueryBuilderAutoSuggestionSearch Component Display different placeholder on IN operators: Expect auto-suggestion field to be displayed with a different placeholder. 2`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <input
        class="form-control"
        placeholder="queryBuilder_value_in_placeholder"
        type="text"
        value="test"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderAutoSuggestionSearch QueryBuilderAutoSuggestionSearch Component Display field if match type is equals: Expect date field not to be displayed since the match type is not equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderAutoSuggestionSearch QueryBuilderAutoSuggestionSearch Component Display field if match type is equals: Expect date field to be displayed since match type is equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <style>
        
          .autosuggest-container {
            position: relative;
            width: 100%;
          }
          .autosuggest .suggestions-container {
            display: none;
          }
          .autosuggest .suggestions-container-open {
            display: block;
            position: absolute;
            width: 100%;
            z-index: 20;
          }
          .autosuggest .suggestion-highlighted { 
            background-color: #ddd;
            cursor: pointer;
          }
        
      </style>
      <div
        class="autosuggest"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-1"
          class="autosuggest-container"
          role="combobox"
        >
          <input
            aria-autocomplete="list"
            aria-controls="react-autowhatever-1"
            autocomplete="none"
            class="form-control"
            placeholder="Type here to search."
            type="text"
            value="test"
          />
          <div
            class="suggestions-container"
            id="react-autowhatever-1"
            role="listbox"
          />
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
